src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -891,6 +893,105 @@ private unsafe void ReadItemsFromSource(ref UnmanagedWriteBuffer writeBuffer, in
+        {
+            private readonly DocumentsOperationContext _context;
+            private readonly long _maxSizeForContextUseInBytes;
+            private readonly long _minSizeToAllocateNonContextUseInBytes;
+            public long TotalDocumentsSizeInBytes { get; private set; }
+
+            private List<Allocation> _nativeAllocationList;
+            private Allocation _currentAllocation;
+
+            public IncomingReplicationAllocator(DocumentsOperationContext context, DocumentDatabase database)
+            {
+                _context = context;
+
+                var maxSizeForContextUse = database.Configuration.Replication.MaxSizeToSend * 2 ??
+                              new Size(128, SizeUnit.Megabytes);
+
+                _maxSizeForContextUseInBytes = maxSizeForContextUse.GetValue(SizeUnit.Bytes);
+                var minSizeToNonContextAllocationInMb = PlatformDetails.Is32Bits ? 4 : 16;
+                _minSizeToAllocateNonContextUseInBytes = new Size(minSizeToNonContextAllocationInMb, SizeUnit.Megabytes).GetValue(SizeUnit.Bytes);
+            }
+
+            public byte* AllocateMemoryForDocument(int size)
+            {
+                TotalDocumentsSizeInBytes += size;
+                if (TotalDocumentsSizeInBytes < _maxSizeForContextUseInBytes)
+                {
+                    _context.Allocator.Allocate(size, out var output);
+                    return output.Ptr;
+                }
+
+                if (_currentAllocation == null || _currentAllocation.Free < size)
+                {
+                    // there can be a document that is larger than the minimum
+                    var sizeToAllocate = Math.Max(size, _minSizeToAllocateNonContextUseInBytes);
+
+                    var allocation = new Allocation(sizeToAllocate);
+                    if (_nativeAllocationList == null)
+                        _nativeAllocationList = new List<Allocation>();
+
+                    _nativeAllocationList.Add(allocation);
+                    _currentAllocation = allocation;
+                }
+
+                return _currentAllocation.GetMemory(size);
+            }
+
+            public void Dispose()
+            {
+                if (_nativeAllocationList == null)
+                    return;
+
+                foreach (var allocation in _nativeAllocationList)
+                {
+                    allocation.Dispose();
+                }
+            }
+
+            private class Allocation : IDisposable
+            {
+                private readonly byte* _ptr;
+                private readonly long _allocationSize;
+                private readonly NativeMemory.ThreadStats _threadStats;
+                private long _used;
+                public long Free => _allocationSize - _used;
+
+                public Allocation(long allocationSize)
+                {
+                    _ptr = NativeMemory.AllocateMemory(allocationSize, out var threadStats);
```

Comment:
I'm not sure we want to use not pooled memory here always.
Maybe we better store some usage statistics and change this to some pooled solution, if we have a pathological high memory usage during replication

Reply:
This only kicks-in after (assuming default configuration) 128mb